realm/src/androidTest/java/io/realm/RealmInMemoryTest.java

```diff
@@ -0,0 +1,242 @@
+        onDiskRealm = Realm.getInstance(encConf);
+        assertEquals(onDiskRealm.allObjects(Dog.class).size(), 1);
+        onDiskRealm.close();
+        // Test with a wrong key to see if it fails as expected.
+        try {
+            RealmConfiguration wrongKeyConf = new RealmConfiguration.Builder(getContext())
+                    .name(encFileName)
+                    .encryptionKey(TestHelper.getRandomKey(42))
+                    .build();
+            Realm.getInstance(wrongKeyConf);
+            fail("Realm.getInstance should fail with illegal argument");
+        } catch (IllegalArgumentException ignored) {
+        }
+    }
+
+    // Test below scenario:
+    // 1. Create a in-memory Realm instance in the main thread.
+    // 2. Create a in-memory Realm with same name in another thread.
+    // 3. Close the in-memory Realm instance in the main thread and the Realm data should not be released since
+    //    another instance is still held by the other thread.
+    // 4. Close the in-memory Realm instance and the Realm data should be released since no more instance with the
+    //    specific name exists.
+    public void testMultiThread() throws InterruptedException, ExecutionException {
+        final AtomicBoolean isWorkerThreadReady = new AtomicBoolean(false);
+        final AtomicBoolean isWorkerThreadFinished = new AtomicBoolean(false);
+        final AtomicBoolean isRealmInMainThreadClosed = new AtomicBoolean(false);
+
+        // Step 2.
+        Thread workerThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(inMemConf);
+                realm.beginTransaction();
+                Dog dog = realm.createObject(Dog.class);
+                dog.setName("DinoDog");
+                realm.commitTransaction();
+                isWorkerThreadReady.set(true);
+
+                assertEquals(realm.allObjects(Dog.class).size(), 1);
+
+                // Wait until Realm instance closed in main thread
+                try {
+                    while (!isRealmInMainThreadClosed.get()) {
+                        Thread.sleep(5);
+                    }
+                } catch (InterruptedException e) {
+                    fail("Worker thread was interrupted.");
+                }
+
+                realm.close();
+                isWorkerThreadFinished.set(true);
+            }
+        });
+        workerThread.start();
+
+
+        // Wait until the worker thread started
+        while (!isWorkerThreadReady.get()) {
+            Thread.sleep(5);
+        }
+
+        // refresh will be ran in the next loop, manually refresh it here.
+        testRealm.refresh();
+        assertEquals(testRealm.allObjects(Dog.class).size(), 1);
+
+        // Step 3.
+        // Release the main thread Realm reference, and the worker thread hold the reference still
+        testRealm.close();
+        // Step 4.
+        isRealmInMainThreadClosed.set(true);
```

Comment:
Hey, you are right! It could be 0 or 1 since no waiting lock for the `realm.close()`. I will move it before the `isRealmInMainThreadClosed.set(true);`.